### ISSUE TITLE
Parser's buffer is not emptied upon reconnect

### ISSUE THREAD

**Issue by arjenm** [stomp-php/stomp-php#40]
After encountering the just reported 0-byte issue https://github.com/stomp-php/stomp-php/issues/39, I also noticed something else that increased the impact of that problem.

When we get a ConnectionException (for instance from Connection::readFrame), we do a disconnect+connect to attempt a reset of the connection.

But the Connection's Parser may have data (via the addData-calls) in its buffer-property.

If that is the case, the first message in its buffer will be whatever that was, rather than the just received CONNECTED-message. So in short:

```
function someReadMethod()
{
  try
  {
     $frame = $stomp->readFrame();
     // There is a read-error after having read some data in readFrame, 
     // so the catch below is triggered
  }
  catch(ConnectionException $e)
  {
    $stomp->disconnect();
    $stomp->connect();

   // At this point, the stomp-connection is not connected due to the data in the $parser->buffer
   // Which contained (a part of) some other response
  }
}
```

Whenever a new connection is used, that buffer should probably be empty initially . I'm not sure at what point that buffer should be reset, probably at connect-time.

Perhaps the cleanest way to fix this, is to create a new Parser-property in the Connection whenever Connection::connect() is creating a new socket. Rather than during the constructor of the Connection.
Although it may loose some of its settings (like the legacyMode).

This issue also exists in both 3.x and master (as far as I can tell from the code)

**Comment by staabm** [comment]
would be great if you could provide a unit test which covers this scenario.

**Comment by arjenm** [comment]
That would require to overwrite the output of php's built in stream_get_line, do you have a preferred method for such unit tests?

As far as I can see there are a few possible scenario's:
- by putting the unit test in a namespace and creating a namespaced version of that method (i.e. effectively Stomp\stream_get_line() )
- with runkit

Both sound somewhat complicated, do you have a preference?

The way I noticed this, is via a xdebug function trace. The buffer simply start with a partially read MESSAGE-frame and that message was not yet deemed completely read. The CONNECTED-frame was appended to the end of the buffer.

**Comment by staabm** [comment]
I guess we only need to test whether $stomp->connect() leaves a empty buffer in case it was filled before connect()

**Comment by arjenm** [comment]
Actually a non-empty buffer with a offset equal to the size of the buffer should be ok as well, but emptied (and both offset and bufferSize reset to 0) is probably a bit safer :)

It is actually hard to trigger, in almost all scenario's that buffer should be empty.

But you can deliberately fill the Parser's buffer via reflection and do a connect, and see what happens with the Parser's buffer.

**Comment by staabm** [comment]
You could also inject a mock and see whether reset is called from the outside

**Comment by arjenm** [comment]
Sure, but that requires the actual existence of such a reset-function ;)

**Comment by jmglsn** [comment]
I just spend some time to create a test for #39, was fun too. But it could help you, nor?

From my point of view we should offer a reset function within the parser. The reason: if there is any data that would be varnished, we should trigger another exception.

**Comment by jmglsn** [comment]
Just added a method `Parser::flushBuffer()`. Parser is exposed by `Connection::getParser()`. 
Should give you anything you need to trigger a clean reconnect.

Will be included in `4.0.1` and `3.0.6`.
